File: LC_54_Spiral_Matrix.py
#TC-> O(m*n)
#SC->O(1)

class Solution(object):
    def spiralOrder(self, matrix):
        """
        :type matrix: List[List[int]]
        :rtype: List[int]
        """
        if(matrix == None or len(matrix) == 0 or len(matrix[0])==0 ):
            return []
        m = len(matrix) #r
        n = len(matrix[0])
        result = []
        #row wise
        top = 0
        bottom = m-1
        #column wise
        left = 0
        right = n-1

        while(top<=bottom and left<=right):
            #top row filling
            for j in range(left, right+1): #fills 1,2,3
                result.append(matrix[top][j])
            top = top + 1

            #right columns filling
            for i in range(top,bottom+1):#fills 6,9 : remember it fills from 6 and not 3, top is incremented
                result.append(matrix[i][right])
            right = right - 1

            #bottom row filling
            if top<=bottom:
                for j in range(right, left-1, -1):#fills 8,7
                    result.append(matrix[bottom][j])
                bottom = bottom - 1

            #left column filling
            if left<=right:
                for i in range(bottom, top-1, -1):#fills 4
                    result.append(matrix[i][left])
                left = left + 1 

        return result

File: LC54-spiralMatrix.java

// time=  O(m*n) , Space = O(1)
import java.util.*;
class Solution {
    public List<Integer> spiralOrder(int[][] matrix) {
        
        int row = matrix.length;
        int col = matrix[0].length;

        List<Integer> res = new ArrayList<>();

        int top =0;
        int bot =row-1;
        int left =0;
        int right =col-1;
        
        while (top<=bot && left <= right) {
            
            // top row, left to right
            for (int i=left; i<=right; i++){
                res.add(matrix[top][i]);
            }
            top++;

            // right col; top to bottom
            for (int i=top;i<=bot; i++){
                res.add(matrix[i][right]);
            }
            right--;

            // bot row, right to left, if top <= bot
            if (top <= bot) {
                for (int i=right;i>=left; i--) {
                    res.add(matrix[bot][i]);
                }
            }
            bot--;

            // left col, bot to top, if top >= bot
            if (left <= right) {
                for (int i=bot;i>=top; i--) {
                    res.add(matrix[i][left]);
                }
            }
            left++;

        }
        return res;
    }
}